larq_zoo/birealnet.py

```diff
@@ -1,172 +1,188 @@
+        out = tf.keras.layers.Conv2D(
+            self.filters,
+            (7, 7),
+            strides=2,
+            kernel_initializer=self.kernel_initializer,
+            padding="same",
+            use_bias=False,
+        )(img_input)
+        out = tf.keras.layers.BatchNormalization(momentum=0.8)(out)
+        out = tf.keras.layers.MaxPool2D((3, 3), strides=2, padding="same")(out)
 
+        # Layer 2
+        out = self.residual_block(out, filters=self.filters)
 
-@registry.register_hparams(birealnet)
-class default(HParams):
-    filters = 64
-    learning_rate = 5e-3
-    decay_schedule = "linear"
-    epochs = 300
-    batch_size = 512
-    input_quantizer = "approx_sign"
-    kernel_quantizer = "magnitude_aware_sign"
-    kernel_constraint = "weight_clip"
-    kernel_initializer = "glorot_normal"
+        # Layer 3 - 5
+        for _ in range(3):
+            out = self.residual_block(out)
 
-    @property
-    def optimizer(self):
-        if self.decay_schedule == "linear_cosine":
-            lr = tf.keras.experimental.LinearCosineDecay(self.learning_rate, 750684)
-        elif self.decay_schedule == "linear":
-            lr = tf.keras.optimizers.schedules.PolynomialDecay(
-                self.learning_rate, 750684, end_learning_rate=0, power=1.0
-            )
-        else:
-            lr = self.learning_rate
-        return tf.keras.optimizers.Adam(lr)
+        # Layer 6 - 17
+        for _ in range(3):
+            out = self.residual_block(out, double_filters=True)
+            for _ in range(3):
+                out = self.residual_block(out)
+
+        # Layer 18
+        if self.include_top:
+            out = tf.keras.layers.GlobalAvgPool2D()(out)
+            out = tf.keras.layers.Dense(self.num_classes, activation="softmax")(out)
+
+        model = tf.keras.Model(inputs=img_input, outputs=out, name="birealnet18")
+
+        # Load weights.
+        if self.weights == "imagenet":
+            # Download appropriate file
+            if self.include_top:
+                weights_path = utils.download_pretrained_model(
+                    model="birealnet",
+                    version="v0.3.0",
+                    file="birealnet_weights.h5",
+                    file_hash="6e6efac1584fcd60dd024198c87f42eb53b5ec719a5ca1f527e1fe7e8b997117",
+                )
+            else:
+                weights_path = utils.download_pretrained_model(
+                    model="birealnet",
+                    version="v0.3.0",
+                    file="birealnet_weights_notop.h5",
+                    file_hash="5148b61c0c2a1094bdef811f68bf4957d5ba5f83ad26437b7a4a6855441ab46b",
+                )
+            model.load_weights(weights_path)
+        elif self.weights is not None:
+            model.load_weights(self.weights)
+        return model
 
 
 def BiRealNet(
-    include_top=True,
-    weights="imagenet",
-    input_tensor=None,
-    input_shape=None,
-    classes=1000,
-):
+    *,  # Keyword arguments only
```

Comment:
If you have `def fn(a, b, *, c=5, d=3.14): ...` then basically you can't pass `c` or `d` as positional arguments, so `fn(1, 2, 3, 4)` would fail; you would have to do `fn(1, 2, c=3, d=4)`.

Reply:
It's a good way to force people to be explicit when calling the function to reduce accidental errors, e.g. TF uses this so that you can't accidentally swap `logits` and `labels` when calling their various loss functions.

Reply:
Intresting I did not know this yet.